src/text/line.rs

```diff
@@ -586,35 +555,86 @@ impl WidgetRef for Line<'_> {
-                width: span_width.min(area.right().saturating_sub(x)),
-                ..area
-            };
-            span.render(span_area, buf);
-            x = x.saturating_add(span_width);
-            if x >= area.right() {
-                break;
-            }
+    }
+}
+
+/// Renders all the spans of the line that are after the `span_offset` in the line and that are
+/// visible in the given area (starting rendering at `area.x + buf_x_offset`). `buf_x_offset`
+/// probably should just be used to modify area.x instead of being a parameter, but it's here
+/// because that would have made the above code just a little uglier
+///
+/// Algorithm:
+/// 1. For each span in the line:
+///    1. If the span is completely before the `span_offset`, skip it
+///    2. If the span is partially before the `span_offset`, render the right side of the span
+///    3. If the span is completely after the `span_offset`, render the whole span
+/// 2. Indent `area` to account for the width of the rendered span
+/// 3. Update `span_offset` to 0 after the first span is rendered in full
+/// 4. Repeat
+fn render_spans(spans: &[Span], mut area: Rect, buf: &mut Buffer, mut span_offset: u16) {
+    for span in spans {
+        let span_width = span.width() as u16;
+        // ignore spans that are completely before the offset
+        if span_offset >= span_width {
+            span_offset -= span_width;
+            continue;
+        }
+        if area.is_empty() {
+            break;
+        }
+        if span_offset > 0 {
+            // the first span is only partially visible, so truncate the start and render it
+            // starting at span offset
+            let render_width = span_width - span_offset;
+            let (content, display_width) =
+                span.content.unicode_truncate_start(render_width as usize);
+            let display_width = display_width as u16;
+
+            // if the truncation has truncated an initial grapheme, then we need to start rendering
+            // from a position that takes that into account by indenting the start of the area
+            area = indent(area, render_width - display_width);
+            let right = Span::styled(content, span.style);
+            right.render_ref(area, buf);
+
+            area = indent(area, display_width);
+            span_offset = 0; // ensure that the next span is rendered in full
+        } else {
+            // render the whole span
+            span.render_ref(area, buf);
+            area = indent(area, span_width);
         }
     }
 }
 
+/// indents a given `area`'s x value by the given `x_offset`, returning a new `Rect`
+///
+/// TODO: this should probably be move into a pubic method on `Rect` as it's a common operation
+/// that's used in a few places
+const fn indent(area: Rect, x_offset: u16) -> Rect {
+    Rect {
+        x: area.x.saturating_add(x_offset),
+        width: area.width.saturating_sub(x_offset),
+        ..area
+    }
+}
+
```

Comment:
Should we add functions called `shift_x` and `shift_y` that take signed integers? This is functionality that I've wanted as well.

Reply:
Yes - my two problems with introducing this are:
- what to name it so that the name obviously evokes that we're increasing x, while decreasing the width
- should it be able to work on all sides (left, right, top, bottom)
- should it be a single method or multiple
- counting things points and not updating the number of points I'm making

Reply:
It might be possible to express it as a single method:

```rust
rect.shift(Corner::TopLeft, (-1, 5))
```

Alternatives:

- `rect.move(::Corner, (x, y))`
- `rect.drag(::Corner, (x, y))` (I'm picturing click and drag the corner, although this might not be intuitive)

Reply:
I don't like either name
- shift implies moving without changing the size
- move is a reserved keyword
- drag is a bad name because it implies an event that we're not actually handliing

"Indent" is the best I have for now, or perhaps some other verb that means to modify / change the size in some way.

Reply:
Either way - not something to fix in this PR - this method is private and useful here. We can iterate without breaking on what the public version of that looks like.

Reply:
Rust has a [`AddAssign`](https://doc.rust-lang.org/std/ops/trait.AddAssign.html) trait which has the `add_assign` method. (or `sub_assign`, `mul_assign` and so on)
Doesnt seem to improve the naming…

`indent_x` would be a good name for it for my taste. Not sure whether there needs to be indent from the right. `indent_y` might be neat. But yeah, not part of this PR to stabilize this.